### PR TITLE
Hide unimportant details from service details and show some important ones instead

### PR DESCRIPTION
**Pull request by KKoukiou**
So I was thinking how I can solve https://github.com/cockpit-project/cockpit/issues/14321 
Went to service details and I realized that so much information there are not useful for most scenarios, or at least they were never useful for me till now. 
So I though I should hide them by default, and display indead some more useful information, like the port a socket activated service is exposed at, for example.

Having the above mentioned property exposed we can move forward in the future with linking the service details of service sockets to the firewall page. 

What do you think? @marusak ?

With my changes cockpit.socket looks like this:
Screen Shot 2021-04-16 at 18 50 18

### CODE REVIEW

**Comment by garrett** [comment]
Ooooo, I :heart: this! You've got my approval for this PR when it's time for it.

Great work!

File: pkg/systemd/services/service-details.jsx
Comment (marusak) [inline pkg/systemd/services/service-details.jsx:486]: Is this how much it currently consumes? If so, is it dynamical changing over time?

File: pkg/systemd/services/service-details.jsx
Comment (marusak) [inline pkg/systemd/services/service-details.jsx:455]: This commit makes previous one about now showing this item obsolete.

**Comment by marusak** [comment]
I like it a lot! Especially since this is the whole screen on my laptop screen:
screenbefore
And this is with this PR:
afterrel

The space is used just soo much better. Also having `Listen` is very useful.

It shows some random `0` though, can you please take a look at that?
justzero


Unrelated but noticed that there is not bottom padding, I cannot scroll any further than:
anyfrther
It is something we should take a look at.

File: pkg/systemd/services/services.jsx
Comment (martinpitt) [inline pkg/systemd/services/services.jsx:635]: This looks like a copy&paste error.

**Review by martinpitt — CHANGES_REQUESTED** [review]
So cool, thanks @KKoukiou !

File: pkg/systemd/services/service-details.jsx
Comment (KKoukiou) [reply pkg/systemd/services/service-details.jsx:486]: It is but it's also not updated asynchronously right now, let me see how I can get an event for that.

Very nice catch ;)

Comment (KKoukiou) [reply pkg/systemd/services/service-details.jsx:486]: So this seems a bit hard, according to systemd documentation this property never emits a signal when changed. https://man7.org/linux/man-pages/man5/org.freedesktop.systemd1.5.html

Comment (martinpitt) [reply pkg/systemd/services/service-details.jsx:486]: Then I suppose it could update itself every 30s or so through a timer?

File: pkg/systemd/services/services.jsx
Comment (KKoukiou) [reply pkg/systemd/services/services.jsx]: Fixed.

File: pkg/systemd/services/service-details.jsx
Comment (KKoukiou) [reply pkg/systemd/services/service-details.jsx:455]: Removed.

File: pkg/systemd/services/service-details.jsx
Comment (KKoukiou) [reply pkg/systemd/services/service-details.jsx]: Ok polling it is.

File: pkg/systemd/services/service-details.jsx
Comment (martinpitt) [inline pkg/systemd/services/service-details.jsx:212]: This is a bit weird as state -- should rather be a const above the class? `METRICS_POLL_DELAY` or so?

File: pkg/systemd/services/service-details.jsx
Comment (martinpitt) [inline pkg/systemd/services/service-details.jsx:220]: Not necessary -- these are always called directly, not as a "function pointer" to component properties.

File: pkg/systemd/services/service-details.jsx
Comment (martinpitt) [inline pkg/systemd/services/service-details.jsx:228]: Does this really have to repeat the code in componentDidUpdate()? I thought the latter would be called on initial mounting as well?

File: pkg/systemd/services/service-details.jsx
Comment (martinpitt) [inline pkg/systemd/services/service-details.jsx:235]: At this point this.interval is false-y, so this isn't necessary (apparently it does not crash, though)

File: pkg/systemd/services/service-details.jsx
Comment (martinpitt) [inline pkg/systemd/services/service-details.jsx:245]: Does this cause any error message when this.interval is invalid? To be on the safe side, this could explicitly check.

File: pkg/systemd/services/service-details.jsx
Comment (martinpitt) [inline pkg/systemd/services/service-details.jsx:258]: cockpit.format is a bit overkill for this -- simple `+` will do.

File: pkg/systemd/services/service-details.jsx
Comment (martinpitt) [inline pkg/systemd/services/service-details.jsx:262]: It makes sense that `MemoryCurrent` does not generate PropertiesChanged signals, but I'm a bit surprised that changes to `Listen` don't? Is it even realistic that `Listen` changes *while* a service is active? If we only need to poll for MemoryCurrent, then GetAll could be reduced to a much cheaper direct Get.

File: pkg/systemd/services/service-details.jsx
Comment (martinpitt) [inline pkg/systemd/services/service-details.jsx:264]: Let's please handle errors from GetAll here (with a simple console.log) to avoid a page Oops in weird cases.

File: pkg/systemd/services/service-details.jsx
Comment (martinpitt) [inline pkg/systemd/services/service-details.jsx:272]: This feels a bit too verbose, but harmless enough.

File: test/verify/check-system-services
Comment (martinpitt) [inline test/verify/check-system-services:274]: Please 
 - check that there is no Listen
 - stopping the unit makes Memory go away, starting it again brings it back
 - ideally it should also test that the regular update works. Schema: Let test.service wait for a /tmp/do-memhog file, then have it allocate a big chunk (see check-metrics for existing code); the test then gets to this point, creates do-memhog, and ensures that the shown value actually increases (it should take at most 30s, so our standard timeouts seem fine)

File: test/verify/check-system-services
Comment (martinpitt) [inline test/verify/check-system-services:340]: ... and does not show Memory

File: pkg/systemd/services/service-details.jsx
Comment (martinpitt) [inline pkg/systemd/services/service-details.jsx:462]: Style nitpick: replace `=> { return (..) }` with a more direct `=> (..)` (this is what lambda functions are really meant to be..)

File: pkg/systemd/services/service-details.jsx
Comment (martinpitt) [inline pkg/systemd/services/service-details.jsx:463]: why the split/join dance? keys can contain spaces

File: pkg/systemd/services/service-details.jsx
Comment (martinpitt) [inline pkg/systemd/services/service-details.jsx:467]: I thought we moved away from `disabled` classes and use the proper `disabled` attribute. I know this is just moving existing code, but it is now the *only* place in cockpit that uses a "disabled" class. If there's a good reason for that, we can look at that in a follow-up, otherwise this may be a good time to clean this up?

File: test/verify/check-system-services
Comment (martinpitt) [inline test/verify/check-system-services:533]: double space (Sanne would be *so* proud! :grin:)

File: test/verify/check-system-services
Comment (martinpitt) [inline test/verify/check-system-services:540]: Do these tests cover the trigger relationships somewhere, to make sure they work as intended?

**Review by martinpitt — CHANGES_REQUESTED** [review]
This works and looks really well :clap: Many thanks for this!

File: pkg/systemd/services/service-details.jsx
Comment (martinpitt) [reply pkg/systemd/services/service-details.jsx:262]: In other words, the Listen property is defined by the unit's `ListenStream=` definition. So changing that would require a daemon-reload and a service restart.

File: pkg/systemd/services/service-details.jsx
Comment (KKoukiou) [reply pkg/systemd/services/service-details.jsx:228]: componentDidUpdate method is not called for the initial render. So when we show the page first time we need to start the polling. Then everytime the page updates, this might mean that the unit changed state, so we might need to start or stop the polling.

Comment (martinpitt) [reply pkg/systemd/services/service-details.jsx:228]: Ack, thanks; I suppose I was thinking of `useEffect()` as the thing that runs both on initial mount and on update. (That may still be the more elegant way, but let's not block on this).

File: test/verify/check-system-services
Comment (KKoukiou) [reply test/verify/check-system-services:540]: There are some relationships test, but I will check the trigger relationships explicitly in addition.

File: pkg/systemd/services/service-details.jsx
Comment (KKoukiou) [reply pkg/systemd/services/service-details.jsx:212]: Done

File: pkg/systemd/services/service-details.jsx
Comment (KKoukiou) [reply pkg/systemd/services/service-details.jsx:235]: Removed.

File: pkg/systemd/services/service-details.jsx
Comment (KKoukiou) [reply pkg/systemd/services/service-details.jsx:245]: Seems, not but guarded it anyway.

File: pkg/systemd/services/service-details.jsx
Comment (KKoukiou) [reply pkg/systemd/services/service-details.jsx:463]: It was there before, but it's indeed useless.

File: pkg/systemd/services/service-details.jsx
Comment (KKoukiou) [reply pkg/systemd/services/service-details.jsx:467]: Sure I will add a commit on top/

File: test/verify/check-system-services
Comment (KKoukiou) [reply test/verify/check-system-services:533]: Fixed

File: pkg/systemd/services/service-details.jsx
Comment (KKoukiou) [reply pkg/systemd/services/service-details.jsx]: Fixed

File: pkg/systemd/services/service-details.jsx
Comment (martinpitt) [inline pkg/systemd/services/service-details.jsx:233]: It still amazes me how a socket unit could change its `ListenStream` property while it's running. Did you observe that in practice? How would one do that?

File: pkg/systemd/services/service-details.jsx
Comment (martinpitt) [inline pkg/systemd/services/service-details.jsx:290]: (copying from previous review round, still relevant) cockpit.format is a bit overkill for this -- simple + will do.

File: pkg/systemd/services/service-details.jsx
Comment (martinpitt) [inline pkg/systemd/services/service-details.jsx]: but... why? What is wrong with `key={rel.Name}`? The spaces don't hurt at all in a key. And as split() is a no-op here, it obviously worked before :grin:

**Review by martinpitt — CHANGES_REQUESTED** [review]
Thanks! There are still quite a number of issues from the previous review round, I marked the others as resolved. Many of the middle ones are unhelpfully hidden by github, you have to expand them :crying_cat_face:

File: pkg/systemd/services/service-details.jsx
Comment (KKoukiou) [reply pkg/systemd/services/service-details.jsx:233]: No I did not, I just saw it in the API documentation that it emits signal when changed. Also in your previous review I thought you wanted me to write a event listener to follow changes for this property, but I guess I misunderstood the comment. Shall I just remove it then? I also feel it's useless to have an even listener for this.

File: pkg/systemd/services/service-details.jsx
Comment (KKoukiou) [reply pkg/systemd/services/service-details.jsx]: This is implemented now: services: replace custom disabled class in service relationships comma list

File: test/verify/check-system-services
Comment (KKoukiou) [reply test/verify/check-system-services]: there is already a test for it: see string:

self.assertIn(b.text("#Triggers .pf-c-description-list__text"), ["dbus.service", "dbus-broker.service"])

File: pkg/systemd/services/service-details.jsx
Comment (KKoukiou) [reply pkg/systemd/services/service-details.jsx]: So I can remove the event listener I have, but in that case, I still need to call the Get API for Listen in componentDidUpdate or else move the fetching of Listen in the main services.jsx file, which actually refetches some unit properties on daemon reload or unit events. I would rather do the later.

File: pkg/systemd/services/service-details.jsx
Comment (martinpitt) [reply pkg/systemd/services/service-details.jsx:233]: Sorry for the misunderstanding! Yes, please remove it all; it can't change while a .socket unit is running.

File: test/verify/check-system-services
Comment (KKoukiou) [reply test/verify/check-system-services:341]: Socket units also have CurrentMemory shown in the UI

Comment (martinpitt) [reply test/verify/check-system-services:341]: Oh wow -- that shouldn't ever be something else than "trivial" :grin:

Comment (KKoukiou) [reply test/verify/check-system-services:341]: Indeed. I could also hide it, but systemctl status shows it as well, so..

**Comment by KKoukiou** [comment]
Still some test failures, marking as draft till these are resolved.

File: pkg/systemd/services/service-details.jsx
Comment (KKoukiou) [reply pkg/systemd/services/service-details.jsx]: So I moved the 'LIsten' property fetching one component up, because where it was we would not receive updates even if the unit restarted.

**Comment by martinpitt** [comment]
test-static-code [fails](https://github.com/cockpit-project/cockpit/pull/15701/checks?check_run_id=2417239742):
```
test/verify/check-system-services:22: 'time' imported but unused
test/verify/check-system-services:810: redefinition of unused 'time' from line 22
```

**Comment by KKoukiou** [comment]
> test-static-code [fails](https://github.com/cockpit-project/cockpit/pull/15701/checks?check_run_id=2417239742):
> 
> ```
> test/verify/check-system-services:22: 'time' imported but unused
> test/verify/check-system-services:810: redefinition of unused 'time' from line 22
> ```

Ups sorry, reposted.

File: pkg/systemd/services/services.jsx
Comment (martinpitt) [inline pkg/systemd/services/services.jsx:618]: `.endsWith("socket")` works without magic numbers.

File: pkg/systemd/services/service-details.jsx
Comment (martinpitt) [inline pkg/systemd/services/service-details.jsx:264]: Would be nice to log the actual exception.

File: test/verify/check-system-services
Comment (martinpitt) [inline test/verify/check-system-services]: `wait_in_text("#listen", "/run/dbus/")` ? to make sure it doesn't show something empty or bogus?

File: test/verify/check-system-services
Comment (martinpitt) [inline test/verify/check-system-services:407]: No need for this bit

File: test/verify/check-system-services
Comment (martinpitt) [inline test/verify/check-system-services:421]: That would fail if the JS code's interval timer happened to start at the same time as the test -- which is actually fairly likely. Let's just use the standard timeout here.

**Review by martinpitt — CHANGES_REQUESTED** [review]
This works really nicely now :+1: Also thanks for the nice dynamic memory test case!

A few remaining easy nitpicks.

File: test/verify/check-system-services
Comment (KKoukiou) [reply test/verify/check-system-services]: I have tihs self.assertIn("Stream", b.text("#listen .pf-c-description-list__text")), what you commented on seems outdated version.

Comment (KKoukiou) [reply test/verify/check-system-services]: I can add the whole thing, /run/dbus (Stream) though, I was just trying to avoid distribution specific differences.

Comment (martinpitt) [reply test/verify/check-system-services]: I reviewed the commits one by one indeed; `/run/` isn't distribution specific, that should be perfectly safe. "dbus" certainly ought to appear somewhere as well.

File: pkg/systemd/services/service-details.jsx
Comment (martinpitt) [inline pkg/systemd/services/service-details.jsx:264]: FWIW, the previous "Failed to.." string was fine, I meant to add `ex.toString()` as a second argument. But good enough.

**Review by martinpitt — APPROVED** [review]
Thanks! :+1: assuming the tests agree.

**Comment by martinpitt** [comment]
@marusak: [rhel-9-0](https://logs.cockpit-project.org/logs/pull-15701-20210423-124323-ceca3b95-rhel-9-0/log.html) failed in an interesting way:
```
+ rpm -U --force build-results/cockpit-debuginfo-242.63.g64ad0d211-1.wip.el9.x86_64.rpm build-results/cockpit-system-242.63.g64ad0d211-1.wip.el9.noarch.rpm build-results/cockpit-debugsource-242.63.g64ad0d211-1.wip.el9.x86_64.rpm build-results/cockpit-ws-242.63.g64ad0d211-1.wip.el9.x86_64.rpm build-results/cockpit-tests-242.63.g64ad0d211-1.wip.el9.x86_64.rpm build-results/cockpit-bridge-242.63.g64ad0d211-1.wip.el9.x86_64.rpm build-results/cockpit-packagekit-242.63.g64ad0d211-1.wip.el9.noarch.rpm build-results/cockpit-storaged-242.63.g64ad0d211-1.wip.el9.noarch.rpm build-results/cockpit-pcp-242.63.g64ad0d211-1.wip.el9.x86_64.rpm build-results/cockpit-242.63.g64ad0d211-1.wip.el9.x86_64.rpm
error: Failed dependencies:
	(selinux-policy >= 34.1.2 if selinux-policy-targeted) is needed by cockpit-ws-242.63.g64ad0d211-1.wip.el9.x86_64
```

This is some weird collision with your PR? that just landed

**Comment by martinpitt** [comment]
Or perhaps some interaction between https://github.com/cockpit-project/cockpit/pull/15707 and the image refresh in https://github.com/cockpit-project/bots/pull/1955 ? They landed in parallel

**Comment by martinpitt** [comment]
@garrett: I made new screenshots (the original one was way too wide) and added a release note. Can you please have a look?

**Comment by garrett** [comment]
@martinpitt Cool!

I do remember a screenshot showing a port; these don't though?

**Comment by garrett** [comment]
Here's a cropped screenshot from the above thread that shows a port:

114979796-c81e3e00-9e8b-11eb-9b3f-681fcb9035cb-3

**Comment by martinpitt** [comment]
@garrett Right, I picked an Unix socket instead of a TCP one, but a TCP socket is fine as well. I just didn't want to use Cockpit's own service, as that felt a little too self-referential. But fine for me, of course :) Do you want to do the corresponding service one as well, or want me to re-do mine?

**Pull request merged**
